codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -39,22 +39,30 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.HashSet;
-import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListSet;
 
 /**
  * Simple implementation of {@link Http2Connection}.
  */
 public class DefaultHttp2Connection implements Http2Connection {
 
+    private static final Comparator<Http2Stream> STREAMID_COMPARATOR = new Comparator<Http2Stream>() {
+        @Override
+        public int compare(Http2Stream o1, Http2Stream o2) {
+            return o1.id() - o2.id();
+        }
+    };
+
     private final Set<Listener> listeners = new HashSet<Listener>(4);
     private final IntObjectMap<Http2Stream> streamMap = new IntObjectHashMap<Http2Stream>();
     private final ConnectionStream connectionStream = new ConnectionStream();
-    private final Set<Http2Stream> activeStreams = new LinkedHashSet<Http2Stream>();
+    private final Set<Http2Stream> activeStreams = new ConcurrentSkipListSet<Http2Stream>(STREAMID_COMPARATOR);
```

Comment:
@Scottmitch 

While I do think the current solution is fine and I don't have a good mental model or any benchmarks about the performance impact of this change, I am open to developing a better solution and/or rethinking the whole interface. How about we open an issue for this where we can discuss this? The problem right now is really time. I am quite busy with other things and my server crashes within 20 seconds after start up because of this error and so I am very interested in getting a fix into Netty asap :-).